tools/codegen/dest/lazy_ir.py

```diff
@@ -0,0 +1,138 @@
+                else:
+                    types.append(NamedCType(binds, ConstRefCType(OptionalCType(BaseCType(valueT)))))
+                    value_types.append(NamedCType(binds, ConstRefCType(OptionalCType(BaseCType(valueT)))))
+            else:
+                types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+                scalar_types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+        elif isinstance(t, ListType):
+            if str(t.elem) == 'Tensor':
+                types.append(NamedCType(binds, BaseCType(valueListT)))
+                value_types.append(NamedCType(binds, BaseCType(valueListT)))
+            elif str(t.elem) == 'Tensor?':
+                types.append(NamedCType(binds, ConstRefCType(ListCType(OptionalCType(BaseCType(valueT))))))
+                value_types.append(NamedCType(binds, ConstRefCType(ListCType(OptionalCType(BaseCType(valueT))))))
+            else:
+                types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+                scalar_types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+        else:
+            raise AssertionError(f"unrecognized type {repr(t)}")
+    return types, value_types, scalar_types
+
+@dataclass(frozen=True)
+class LazyIR:
+    backend_index: BackendIndex
+
+    target: Union[
+        Literal[Target.DEFINITION],
+        Literal[Target.DECLARATION],
+    ]
+
+    # TODO(whc) probably use selector instead of building a separate index for ops to codegen
+    # Selector object to determine which operators to generate
+    # registration code for.
+    selector: SelectiveBuilder
+
+    # The namespace that the kernels are written in. This is just `at::native` for in-tree kernels.
+    cpp_namespace: str
+
+    @method_with_native_function
+    def __call__(self, f: Union[NativeFunctionsGroup, NativeFunction]) -> List[str]:
+        func = f.functional.func if isinstance(f, NativeFunctionsGroup) else f.func
+        if func.name in self.backend_index.index:
+            return self.gen(f)
+        else:
+            return []
+
+    def gen(self, f: Union[NativeFunctionsGroup, NativeFunction]) -> List[str]:
+        # for now, we just want one IR class decl and soon after also the method defs
+        # and we use the functional version not out/inplace.
+        func = f.functional.func if isinstance(f, NativeFunctionsGroup) else f.func
+        class_name = str(func.name).lower().capitalize()
+
+        all_types, value_types, scalar_types = process_ir_types(func)
+        node_ctor_args = ", ".join([f"{i.cpp_type()} {i.name}" for i in all_types])
+        scalar_initializers = ",\n        ".join([f"{t.name}_({t.name})" for t in scalar_types])
+        scalar_decls = ";\n  ".join([f"{t.cpp_type()} {t.name}_" for t in scalar_types])
+        scalar_hashes = ", ".join([f.name for f in scalar_types])
```

Comment:
This is definitely one of those places where, if we just changed the hashing convention, the codegen here would be so much simpler

Reply:
What would be your proposed convention?  Ironically this version you commented on looks fairly simple compared to the nonsense I cobbled together to make things complile.  Taking 'optionals' into account for hashing turns out to be annoying.  Should I just make it so the native 'Hash' function accepts optionals natively and just doesn't include a contribution from them if they are null?

Reply:
I think, if I didn't care about codegen friendliness, I'd probably have each child node compute the *full* hash, and so you just literally pass every argument into a hash function (which is some variadic template thing that successively combines the hashes of each argument). Do need to contribute to hash when optional is null in case of `optional<Tensor>, optional<Tensor>` need to distinguish between `nullopt, t` case and `t, nullopt` case

Reply:
good point. I am already doing a nullopt thing for Value hashes actually. 

But what about scalars?  I hadn't thought about needing to distinguish between say 
optional<float>, optional<bool> as nullopt, false vs 1.0, nullopt.  How do I provide a 'null' value to hash on for scalars?

Reply:
we talked about this in chat; if you don't need cryptographic hash you can just pick some random number like -3333 (like in glibc) for nullopt; for cryptographic hash (no collisions) need to distinguish these cases